src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,262 @@
+        return false;
+    }
+
+    int depth = 0;
+    int tries = 100000;
+    std::vector<std::pair<bool, bool>> selection; // First bool: select the utxo at this index; Second bool: traversing second branch of this utxo
+    selection.assign(utxo_pool.size(), std::pair<bool, bool>(false, false));
+    bool done = false;
+    bool backtrack = false;
+    
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+    
+    // Calculate remaining
+    CAmount remaining = 0;
+    for (CInputCoin utxo : utxo_pool) {
+        remaining += utxo.txout.nValue;
+    }
+    
+    // Depth first search to find 
+    while (!done)
+    {
+        if (tries <= 0) { // Too many tries, exit
+            return false;
+        } else if (value_ret > target_value + cost_of_change) { // Selected value is out of range, go back and try other branch
+            backtrack = true;
+        } else if (value_ret >= target_value) { // Selected value is within range
+            done = true;
+        } else if (depth >= (int)utxo_pool.size()) { // Reached a leaf node, no solution here
+            backtrack = true;
+        } else if (value_ret + remaining < target_value) { // Cannot possibly reach target with amount remaining
+            if (depth == 0) { // At the first utxo, no possible selections, so exit
+                return false;
+            } else {
+                backtrack = true;
+            }
+        } else { // Continue down this branch
+            // Assert that this utxo is not negative. It should never be negative, effective value calculation should have removed it
+            assert(utxo_pool.at(depth).txout.nValue >= 0);
+        
+            // Remove this utxo from the remaining utxo amount
+            remaining -= utxo_pool.at(depth).txout.nValue;
+            // Inclusion branch first (Largest First Exploration)
+            selection.at(depth).first = true;
+            value_ret += utxo_pool.at(depth).txout.nValue;
+            ++depth;
+        }
+
+        // Step back to the previous utxo and try the other branch
+        if (backtrack) {
+            backtrack = false; // Reset
+            --depth;
+
+            // Walk backwards to find the first utxo which has not has its second branch traversed
+            while (selection.at(depth).second) {
+                // Reset this utxo's selection
+                if (selection.at(depth).first) {
+                    value_ret -= utxo_pool.at(depth).txout.nValue;
+                }
+                selection.at(depth).first = false;
+                selection.at(depth).second = false;
+                remaining += utxo_pool.at(depth).txout.nValue;
+
+                // Step back one
+                --depth;
+
+                if (depth < 0) { // We have walked back to the first utxo and no branch is untraversed. No solution, exit.
+                    return false;
+                }
+            }
+            
+            if (!done) {
```

Comment:
This block doesn't happen when `backtrack` is false and `done` is true which happens when a solution is found.